Map Pins will now show on the Azeroth world map.

SinusPi opened the pull request:
Tabs spaced, but hardcoded data kept as-is to have it all in one place.

Commit:
Map Pins will now show on the Azeroth world map.

Nevcairiel commented on HereBeDragons-1.0.lua:
```diff
@@ -308,6 +308,53 @@ if not oldversion or oldversion < 20 then
+                end
+            end
+        end
+        mergeHardcodedData(continentSizes,mapData)
```
Like commented on IRC, it would look nicer if you just add that data into the block a couple lines above that creates the mapData[WORLDMAP_AZEROTH_ID] entry

SinusPi replied:
The way I wrote it keeps the data compact - first the COSMIC and WORLD dummies, then the coordinate filling. To me, it looks more messy when crammed into the data already - but you're the boss here. :)

Nevcairiel commented on HereBeDragons-1.0.lua:
```diff
@@ -308,6 +308,53 @@ if not oldversion or oldversion < 20 then
+                [3] = 53730,
+                [4] = 19600,
+                floors = {
+                    [0] = { 1,1,1,1, "just here to fool some functions that check floors[num]" },
```
Where is this needed? I don't think floors[0] is typically accessed

Nevcairiel commented on HereBeDragons-Pins-1.0.lua:
```diff
@@ -514,7 +519,11 @@ function pins:AddWorldMapIconWorld(ref, icon, instanceID, x, y)
+            currentMapID = 0  -- force 0=Azeroth
+        end
+        local mapData = HBD.mapData[currentMapID]
+        if currentMapID and mapData and (mapData.instance == instanceID or mapData.floors[instanceID]) then
```
Instead of checking mapData.floors[instanceID], I would probabl prefer simply checking for currentMapID == 0, less risk of bad things happening (in the other places that do this as well)

SinusPi replied:
This simply rejects Outland, Draenor and Maelstrom outright - without waiting for HBD:GetZoneCoordinatesFromWorld to return nils. It looks worse, but it's a tiny bit faster. I do agree, though, that the performance gain from skipping just one function call might not be huge.

Nevcairiel commented on HereBeDragons-1.0.lua:
```diff
@@ -308,6 +308,53 @@ if not oldversion or oldversion < 20 then
+                        [2] = 64476,
+                        [3] = 63100,
+                        [4] = 29960
+                    }
```
If these floor data tables would set the "instance" member to the appropriate value, then some extra functions like TranslateZoneCoordinates and GetZoneDistance could actually work with the world map, at least in theory, I didn't test it yet.

SinusPi replied:
Nope, I don't believe they would. The zones are each sized differently, because map artists took liberties. Xlating coords up to the world level and then xlating them back into continents wouldn't preserve the scaling. Thus, there cannot be a "world coordinate system" other than for 0..1 display on the map. Of course, one could jump some hoops and upscale certain rectangular areas of the world before doing the distance calculations, but that would be a complete nightmare...

Nevcairiel commented on HereBeDragons-Pins-1.0.lua:
```diff
@@ -276,7 +276,7 @@ local function UpdateMinimapZoom()
 
 local function PositionWorldMapIcon(icon, data, currentMapID, currentMapFloor)
-    local x, y = HBD:GetZoneCoordinatesFromWorld(data.x, data.y, currentMapID, currentMapFloor)
+    local x, y = HBD:GetZoneCoordinatesFromWorld(data.x, data.y, currentMapID, (currentMapID ~= 0 and currentMapFloor or data.instanceID))
```
This condition should be flipped around, so that a "nil" floor doesn't fall back to sending it the instance id, as that would break things (Lua and/or constructs are not true ternary operations)

ie. "currentMapID == 0 and data.instance or currentMapFloor"

the instance should always be set, so in this direction that behavior should not matter.

SinusPi replied:
Ah, I didn't think currentMapFloor might not be set. Indeed in that case the order matters.

Commit:
Nevcairiel's comments applied.

.floors[0] is no longer needed, ternary-ish and-or structures no longer
risk falling for a missing floor number, and continent data is now in
the main table without a merger function.

Nevcairiel commented:
I did some minor changes, cosmetics and some cleanup, and pushed this into the "world" branch, if you want to do some final testing before I merge it into master eventually.

I did add the instance id for the individual floors which does enable a wider range of functions to work, because they require the instance to be set and the same (ie. this allows directly translating 0..1 coordinates from a zone level to the world level without two intermediate steps)

Nevcairiel commented:
One thing I wondered about, are all continents tested well?
Northrend and Kalimdor having the same width/height in the data makes me suspicious.

SinusPi commented:
Tested as much as they could. The same size surprised me too, but I started by sizing Kalimdor, then I copied the data as a template to the other continents - and Northrend was a good fit already. http://puu.sh/pMvzB/16b671ee20.jpg  Here, continent edges dotted.

SinusPi commented:
So how does this work git-wise, now? I don't seem able to switch to base off your /world branch, I'm still forked off /master, but the changes were merged into /world which I don't see... Should I cancel this pull request or what? I'm not VERY familiar with multi-branch gitting, I'm afraid.

Nevcairiel commented:
I'll just close it when I merge it on the weekend, don't need to do anything. :)

Nevcairiel commented:
Its merged in master now and will also be available as an alpha zip package on wowace soon.
Thanks for your work on this.
